xbmc/filesystem/FileCache.cpp

```diff
@@ -234,8 +252,27 @@ void CFileCache::Process()
     return;
   }
 
+  // How much should we request? Well if we have a front buffer size, use that.
+  // Note however that if more than half of the forward buffer is empty,
+  // that would mean that we're falling behind and if we try to read more
+  // than what is in the buffer now in that situation, we will run out of bytes
+  // before the next read call completes. Hence it never makes sense to read
+  // more than half of the forward cache size at a time as even if that were
+  // to improve throughput, we have too little buffer space to keep up.
+  int64_t requestSize;
+  if (m_forwardCacheSize)
+  {
+    requestSize = m_forwardCacheSize / 2;
+  }
+  else
+  {
+    // We have "unlimited" buffer capabilities, but we still must bound the
+    // request size to something.
+    requestSize = DEFAULT_REQUEST_SIZE;
```

Comment:
I don't think this will work for streaming filesystems like ftp/http/webdav. Afaik the read size there is limited by the internal ringbuffer in CCurlFile. I think the assumption you're doing here is solely based on what smb can do...

Reply:
Yes, only SMB has been tested, as noted.

I can test other things and for sure there's likely to be at least some problem somewhere. Nevertheless I don't see a way to improve performance for the filesystem case without increasing the read size somehow.

Regarding `IFile`s which can't deliver this much data at a time, would it be acceptable to change those so that they return less data in those cases? Roughly matching the behaviour of the posix `read()` syscall?

Reply:
I've quickly tested this with NFS and HTTP on LAN and didn't notice any immediate issues.

Reply:
Ah yes that's probably because we handle big reads in CCurlFile::ReadData(). But still: I think it's a bad idea for at least streamed filesystems, especially since there's not much gain for those doing big reads. 

Wouldn't it be better to simply add something like MaxReadSize() to IFile so filesystems can tell how much they can return at once?

Reply:
That might not be a bad thing to have, but there's quite a few implementations to add it to and I'm not sure a default value can be picked in a reasonable way since it depends on the bitrate. IMO, ideally it should be based on time and the bitrate of the media tuned so we get for example 1-100 read calls per second. I often deal with raw `read()` syscalls so to me letting `Read()` return less than requested feels like second nature to me.

In either case, CCurlFile seems to return short reads sometimes and that currently works (at least before these changes). That doesn't say anything about other types of course but the rest of the code seems to cope with those short reads at least.

I've tried to test the Youtube and TED Talk plugins but those seem broken in master so I didn't get to any `IFile` things that way.

Reply:
@wsnipex Thanks for your testing!

I'm wondering if the NFS code should have `chunkSize = 1` too with similar arguments as for SMBFile. I might dig into that.

Reply:
> That might not be a bad thing to have, but there's quite a few implementations to add it to and I'm not sure a default value can be picked in a reasonable way since it depends on the bitrate. IMO, ideally it should be based on time and the bitrate of the media tuned so we get for example 1-100 read calls per second. I often deal with raw `read()` syscalls so to me letting `Read()` return less than requested feels like second nature.
> 
> In either case, CCurlFile seems to return short reads sometimes and that currently works (at least before these changes). That doesn't say anything about other types of course but the rest of the code seems to cope with those short reads at least.
> 
> I've tried to test the Youtube and TED Talk plugins but those seem broken in master so I didn't get to any `IFile` things that way.

Please keep in mind that, especially with stuff on the internet, when a large read (partially) fails, we need to possibly read a larger block all over again. This is also a reason why you don't want too large reads especially with streaming filesystems. And of course larger reads also require more memory, but that's probably not an issue on most hardware.

I'd make the implementation in a way that it behaves the same as it did before except for the filesystems of which we know they allow larger reads and where it's beneficial, like smb/nfs.

Reply:
libnfs bounds chunk size to whatever the server supports.

Reply:
Normally I would expect the posix `read()` to deal with partial failures by first returning what succeeded and then returning an error or retrying on the next call. Dealing with block boundaries complicate it somewhat and you still have a point about only fixing what's broken. I'll give extending `IFile` a try so we can see what it looks like!

Reply:
Ok, there's an implementation now. It's a little bit hairy in the `CFileCache` class, but overall not too bad I think.